webknossos-jni/src/bucketScanner.cpp

```diff
@@ -78,11 +79,85 @@ JNIEXPORT jlongArray JNICALL Java_com_scalableminds_webknossos_datastore_helpers
+        throwRuntimeException(env, "Native Exception in BucketScanner collectSegmentIds: " + std::string(e.what()));
         return nullptr;
     } catch (...) {
         env->ReleaseByteArrayElements(bucketBytesJavaArray, bucketBytes, 0);
-        throwRuntimeException(env, "Native Exception in BucketScanner");
+        throwRuntimeException(env, "Native Exception in BucketScanner collectSegmentIds");
         return nullptr;
     }
 }
+
+JNIEXPORT jlong JNICALL Java_com_scalableminds_webknossos_datastore_helpers_NativeBucketScanner_countSegmentVoxels
+    (JNIEnv * env, jobject instance, jbyteArray bucketBytesJavaArray, jint bytesPerElement, jboolean isSigned, jlong segmentId) {
+
+    jsize inputLengthBytes = env -> GetArrayLength(bucketBytesJavaArray);
+    jbyte * bucketBytes = env -> GetByteArrayElements(bucketBytesJavaArray, NULL);
+    try {
+
+        const size_t elementCount = getElementCount(inputLengthBytes, bytesPerElement);
+        size_t segmentVoxelCount = 0;
+        for (size_t i = 0; i < elementCount; ++i) {
+            int64_t currentValue = segmentIdAtIndex(bucketBytes, i, bytesPerElement, isSigned);
+            if (currentValue == segmentId) {
+                segmentVoxelCount++;
+            }
+        }
+        env->ReleaseByteArrayElements(bucketBytesJavaArray, bucketBytes, 0);
+        return segmentVoxelCount;
+
+    } catch (const std::exception &e) {
+        env->ReleaseByteArrayElements(bucketBytesJavaArray, bucketBytes, 0);
+        throwRuntimeException(env, "Native Exception in BucketScanner countSegmentVoxels: " + std::string(e.what()));
+        return 0;
+    } catch (...) {
+        env->ReleaseByteArrayElements(bucketBytesJavaArray, bucketBytes, 0);
+        throwRuntimeException(env, "Native Exception in BucketScanner countSegmentVoxels");
+        return 0;
+    }
+}
+
+
+JNIEXPORT jintArray JNICALL Java_com_scalableminds_webknossos_datastore_helpers_NativeBucketScanner_extendSegmentBoundingBox
+    (JNIEnv * env, jobject instance, jbyteArray bucketBytesJavaArray, jint bytesPerElement, jboolean isSigned, jint bucketLength, jlong segmentId,
+      jint bucketTopLeftX, jint bucketTopLeftY, jint bucketTopLeftZ,
+      jint existingBBoxTopLeftX, jint existingBBoxTopLeftY, jint existingBBoxTopLeftZ,
+      jint existingBBoxBottomRightX, jint existingBBoxBottomRightY, jint existingBBoxBottomRightZ) {
+
+    jsize inputLengthBytes = env -> GetArrayLength(bucketBytesJavaArray);
+    jbyte * bucketBytes = env -> GetByteArrayElements(bucketBytesJavaArray, NULL);
+    try {
+
+        const size_t elementCount = getElementCount(inputLengthBytes, bytesPerElement);
+        std::vector<int> bbox = {existingBBoxTopLeftX, existingBBoxTopLeftY, existingBBoxTopLeftZ, existingBBoxBottomRightX, existingBBoxBottomRightY, existingBBoxBottomRightZ};
+        for (int x = 0; x < bucketLength; x++) {
+            for (int y = 0; y < bucketLength; y++) {
+                for (int z = 0; z < bucketLength; z++) {
```

Comment:
Consider iterating over z,y,x to improve cache locality, i.e. in each loop the index increases by 1.